contracts/v0.6/SimpleSwap.sol

```diff
@@ -0,0 +1,178 @@
+    address indexed target,
+    address caller
+  );
+
+  mapping(address => mapping(address => uint256)) private s_swappableAmount;
+
+  /**
+   * @dev deposits tokens from the target of a swap pair but does not return any
+   * @param amount count of liquidity being added
+   * @param source the token that can be swapped for what is being deposited
+   * @param target the token that can is being deposited for swapping
+   */
+  function addLiquidity(
+    uint256 amount,
+    address source,
+    address target
+  )
+    external
+  {
+    _addLiquidity(amount, source, target);
+
+    require(ERC20(target).transferFrom(msg.sender, address(this), amount), "transferFrom failed");
+  }
+
+  /**
+   * @dev withdraws tokens from the target of a swap pair.
+   * Only callable by owner
+   * @param amount count of liquidity being removed
+   * @param source the token that can be swapped for what is being removed
+   * @param target the token that can is being withdrawn from swapping
+   */
+  function removeLiquidity(
+    uint256 amount,
+    address source,
+    address target
+  )
+    external
+    onlyOwner()
+  {
+    _removeLiquidity(amount, source, target);
+
+    require(ERC20(target).transfer(msg.sender, amount), "transfer failed");
+  }
+
+  /**
+   * @dev exchanges the source token for target token
+   * @param amount count of tokens being swapped
+   * @param source the token that is being given
+   * @param target the token that is being taken
+   */
+  function swap(
+    uint256 amount,
+    address source,
+    address target
+  )
+    external
+    nonReentrant()
+  {
+    _removeLiquidity(amount, source, target);
+    _addLiquidity(amount, target, source);
+
+    emit TokensSwapped(amount, source, target, msg.sender);
+
+    require(ERC20(source).transferFrom(msg.sender, address(this), amount), "transferFrom failed");
+    require(ERC20(target).transfer(msg.sender, amount), "transfer failed");
+  }
+
+  /**
+   * @dev send funds that were accidentally transferred back to the owner. This
+   * allows rescuing of funds, and poses no additional risk as the owner could
+   * already withdraw any funds intended to be swapped. WARNING: If not called
+   * correctly this method can throw off the swappable token balances, but that
+   * can be recovered from by transferring the discrepancy back to the swap.
+   * @param amount count of tokens being moved
+   * @param target the token that is being moved
+   */
+  function recoverTransferredTokens(
+    uint256 amount,
+    address target
+  )
+    external
+    onlyOwner()
+  {
+    require(ERC20(target).transfer(msg.sender, amount), "transfer failed");
+  }
+
+  /**
+   * @dev swap tokens in one transaction if the sending token supports ERC677
+   * @param sender address that initially initiated the call to the source token
+   * @param amount count of tokens sent for the swap
+   * @param targetData address of target token encoded as a bytes array
+   */
+  function onTokenTransfer(
```

Comment:
Building off of @kaleofduty's note in `addLiquidity`, we should check the allowlist here as well.

Reply:
FWIW, I think this will be done implicitly thanks to the call to `_addLiquidity`